cmd/metal-api/internal/service/project-service.go

```diff
@@ -161,34 +174,221 @@ func (r projectResource) findProjects(request *restful.Request, response *restfu
+		return
+	}
+
+	var ms metal.Machines
+	err = r.ds.SearchMachines(&datastore.MachineSearchQuery{AllocationProject: &id}, &ms)
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+	if len(ms) > 0 {
+		checkError(request, response, utils.CurrentFuncName(), fmt.Errorf("there are still machines allocated by this project"))
+		return
+	}
+
+	var ns metal.Networks
+	err = r.ds.SearchNetworks(&datastore.NetworkSearchQuery{ProjectID: &id}, &ns)
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+	if len(ns) > 0 {
+		checkError(request, response, utils.CurrentFuncName(), fmt.Errorf("there are still networks allocated by this project"))
+		return
+	}
+
+	var ips metal.IPs
+	err = r.ds.SearchIPs(&datastore.IPSearchQuery{ProjectID: &id}, &ips)
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+
+	if len(ips) > 0 {
+		checkError(request, response, utils.CurrentFuncName(), fmt.Errorf("there are still ips allocated by this project"))
+		return
+	}
+
+	pdr := &mdmv1.ProjectDeleteRequest{
+		Id: p.Project.Meta.Id,
+	}
+	pdresponse, err := r.mdc.Project().Delete(context.Background(), pdr)
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+
+	v1p := mapper.ToV1Project(pdresponse.Project)
+	pcres := &v1.ProjectResponse{
+		Project: *v1p,
+	}
+
+	err = response.WriteHeaderAndEntity(http.StatusOK, pcres)
+	if err != nil {
+		zapup.MustRootLogger().Error("Failed to send response", zap.Error(err))
+		return
+	}
+}
+
+func (r projectResource) updateProject(request *restful.Request, response *restful.Response) {
+	var requestPayload v1.ProjectUpdateRequest
+	err := request.ReadEntity(&requestPayload)
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+	if requestPayload.Project.Meta == nil {
+		err = errors.New("project and project.meta must be specified")
+		checkError(request, response, utils.CurrentFuncName(), err)
+		return
+	}
+
+	existingProject, err := r.mdc.Project().Get(context.Background(), &mdmv1.ProjectGetRequest{Id: requestPayload.Project.Meta.Id})
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+
+	// new data
+	projectUpdateData := mapper.ToMdmV1Project(&requestPayload.Project)
+	// created date is not updateable
+	projectUpdateData.Meta.CreatedTime = existingProject.Project.Meta.CreatedTime
+
+	pur, err := r.mdc.Project().Update(context.TODO(), &mdmv1.ProjectUpdateRequest{
+		Project: projectUpdateData,
+	})
+	if err != nil {
+		checkError(request, response, utils.CurrentFuncName(), err)
+		return
+	}
+
+	v1p := mapper.ToV1Project(pur.Project)
+	err = response.WriteHeaderAndEntity(http.StatusOK, v1.ProjectResponse{
+		Project: *v1p,
+	})
+	if err != nil {
+		zapup.MustRootLogger().Error("Failed to send response", zap.Error(err))
+		return
+	}
+}
+
+func (r projectResource) ipsByProject() (map[string]metal.IPs, error) {
+	ips, err := r.ds.ListIPs()
```

Comment:
Same as for machines (also has `ByProjectID()`)

Reply:
same comment as above

Reply:
use ByProjectID() now